analyses/cell-type-ewings/exploratory_analysis/05-cluster-exploration.Rmd

```diff
@@ -0,0 +1,626 @@
+
+```{r, fig.height=15, fig.width=10}
+# look at clustering results across all params
+ggplot(umap_df, aes(x = UMAP1, y = UMAP2, color = cluster)) +
+  geom_point(alpha = 0.5, size = 0.1) +
+  facet_grid(
+    cols = vars(library_id),
+    rows = vars(nn_param)
+  ) +
+  theme(aspect.ratio = 1)
+```
+
+### Clustering statistics 
+
+Below we calculate a series of statistics: 
+
+- Average silhouette width: This metric evaluates cluster separation. 
+Cells with large positive silhouette widths are closer to other cells in the same cluster than to cells in different clusters. 
+- Average cluster purity: This metric also evaluates cluster separation and tells us the proportion of neighboring cells that are assigned to the same cluster. 
+Higher purity values indicate that clusters are well separated. 
+- Cluster stability: This evaluates how stable the clustering is to input data. 
+Higher values of cluster stability indicate more reproducible clusters. 
+
+```{r, warning=FALSE}
+# get a combined stats dataframe with purity and width for all clusters
+all_stats_df <- sce_list |>
+  purrr::imap(\(sce, library_id){
+    get_cluster_stats(sce, library_id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id")
+```
+
+```{r}
+# summarize the silhouette width and purity for each nn_param used
+summary_stats_df <- all_stats_df |>
+  dplyr::group_by(library_id, nn_param) |>
+  dplyr::summarise(
+    mean_width = mean(width),
+    mean_purity = mean(purity)
+  ) |>
+  dplyr::mutate(
+    # make sure the order is correct
+    nn_param = forcats::fct_relevel(nn_param, "k_5", after = 0)
+  )
+```
+
+
+```{r}
+# silhouette width for different params
+ggplot(summary_stats_df, aes(x = nn_param, y = mean_width)) +
+  geom_point() +
+  facet_wrap(vars(library_id),
+    scales = "free_y"
+  ) +
+  labs(y = "Average silhouette width")
+```
+
+```{r}
+# cluster purity for different params
+ggplot(summary_stats_df, aes(x = nn_param, y = mean_purity)) +
+  geom_point() +
+  facet_wrap(vars(library_id),
+    scales = "free_y"
+  ) +
+  labs(y = "Average cluster purity")
+```
+
+Here we see that both silhouette width and cluster purity peak around k=20 or k=25. 
+For `SCPCL000822`, increasing k after 25 does not change the width or purity, 
+For `SCPCL000824`, increasing k after 20-25 decreases the width and keeps purity around the same. 
+
+```{r, warning=FALSE, message=FALSE}
+# calculate cluster stability
+stability_df <- sce_list |>
+  purrr::imap(\(sce, id){
+    get_cluster_stability(sce, id, all_cluster_results)
+  }) |>
+  dplyr::bind_rows(.id = "library_id") |>
+  dplyr::mutate(
+    k_value = as.factor(k_value),
+    # make sure that k = 5 comes first
+    k_value = forcats::fct_relevel(k_value, "5", after = 0)
+  )
+
+# plot stability across all values of k
+ggplot(stability_df, aes(x = k_value, y = ari)) +
+  geom_point() +
```

Comment:
just so we can make sure we're seeing the points, and I suggested my personal favorite width parameter

```suggestion
  geom_jitter(width = 0.1) +
```